docs/source/scheduling-state.rst

```diff
@@ -30,385 +32,445 @@ system used to perform this tracking.
 :doc:`Scheduling Policies<scheduling-policies>`.
 
+The scheduler keeps internal state about several kinds of entities:
 
-State Variables
----------------
+* Individual tasks known to the scheduler
+* Workers connected to it
+* Clients connected to it
 
-We start with a description of the state that the scheduler keeps on each task.
-Each of the following is a dictionary keyed by task name (described below).
 
-Task description variables
-''''''''''''''''''''''''''
+.. note::
+   Everything listed in the page is an internal detail of how Dask operates.
+   It may change between versions and you should probably avoid relying
+   on it in user code (including on any APIs explained here).
 
-These containers keep task information accross a task's whole lifetime.
 
-* **tasks:** ``{key: task}``:
+Task State
+----------
 
-   Dictionary mapping key to a serialized task.
+Internally, the scheduler moves tasks between a fixed set of states,
+notably ``released``, ``waiting``, ``no-worker``, ``processing``,
+``memory``, ``error``.
 
-   A key is the name of a task, generally formed from the name of the
-   function, followed by a hash of the function and arguments, like
-   ``'inc-ab31c010444977004d656610d2d421ec'``.
+Tasks flow along the following states with the following allowed transitions:
 
-   The value of this dictionary is the task, which is an unevaluated function
-   and arguments.  This is stored in one of two forms:
+.. image:: images/task-state.svg
+    :alt: Dask scheduler task states
 
-   * ``{'function': inc, 'args': (1,), 'kwargs': {}}``; a dictionary with the
-     function, arguments, and keyword arguments (kwargs).  However in the
-     scheduler these are stored serialized, as they were sent from the client,
-     so it looks more like ``{'function': b'\x80\x04\x95\xcb\...', 'args':
-     b'...', }``
+*  *Released*: Known but not actively computing or in memory
+*  *Waiting*: On track to be computed, waiting on dependencies to arrive in
+   memory
+*  *No-worker*: Ready to be computed, but no appropriate worker exists
+   (for example because of resource restrictions, or because no worker is
+   connected at all).
+*  *Processing*: Actively being computed by one or more workers
+*  *Memory*: In memory on one or more workers
+*  *Erred*: Task computation, or one of its dependencies, has encountered an error
+*  *Forgotten* (not actually a state): Task is no longer needed by any client
+   or dependent task
 
-   * ``(inc, 1)``: a tuple satisfying the `dask graph protocol
-     <http://dask.pydata.org/en/latest/graphs.html>`_.  This
-     again is stored serialized.
+In addition to the literal state, though, other information needs to be
+kept and updated about each task.  Individual task state is stored in an
+object named :class:`TaskState` and consists of the following information:
 
-   These are the values that will eventually be sent to a worker when the task
-   is ready to run.
+.. class:: TaskState
```

Comment:
Would it be possible to move this documentation to the class docstring?

Reply:
I'll try to. Hopefully that will render correctly.